clang/test/Driver/sycl-device-optimizations.cpp

```diff
@@ -53,3 +53,59 @@
 // RUN:   %clang -### -fsycl --offload-new-driver %s 2>&1 \
 // RUN:   | FileCheck -check-prefix=CHECK-NO-THRESH %s
 // CHECK-NO-THRESH-NOT: "-mllvm" "-inline-threshold
+
+/// Check that optimizations for sycl device are disabled with -g passed:
+// RUN:   %clang -### -fsycl -g %s 2>&1 \
+// RUN:   | FileCheck -check-prefix=CHECK-DEBUG %s
+// RUN:   %clang_cl -### -fsycl -g %s 2>&1 \
+// RUN:   | FileCheck -check-prefix=CHECK-DEBUG %s
+// CHECK-DEBUG: clang{{.*}} "-fsycl-is-device{{.*}}" "-O0"
+// CHECK-DEBUG: sycl-post-link{{.*}} "-O0"
+// CHECK-DEBUG-NOT: "-O2"
+
+/// Check that optimizations for sycl device are enabled with -g and O2 passed:
+// RUN:   %clang -### -fsycl -O2 -g %s 2>&1 \
+// RUN:   | FileCheck -check-prefix=CHECK-G-O2 %s
+// For clang_cl, -O2 maps to -O3
+// RUN:   %clang_cl -### -fsycl -O2 -g %s 2>&1 \
+// RUN:   | FileCheck -check-prefix=CHECK-G-O3 %s
+// CHECK-G-O2: clang{{.*}} "-fsycl-is-device{{.*}}" "-O2"
+// CHECK-G-O2: sycl-post-link{{.*}} "-O2"
+// CHECK-G-O2-NOT: "-O0"
+// CHECK-G-O3: clang{{.*}} "-fsycl-is-device{{.*}}" "-O3"
+// CHECK-G-O3: sycl-post-link{{.*}} "-O3"
```

Comment:
Side note - the `-O2` to `-O3` behavior is not in xmain, so some adjustments to the test will be needed when this is pulled down.